pkg/telemetry/cassandra_agent/cassandra_agent_config.go

```diff
@@ -0,0 +1,122 @@
+package cassandra_agent
+
+import (
+	"context"
+	"path/filepath"
+
+	cassdcapi "github.com/k8ssandra/cass-operator/apis/cassandra/v1beta1"
+	k8ssandraapi "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
+	telemetryapi "github.com/k8ssandra/k8ssandra-operator/apis/telemetry/v1alpha1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/cassandra"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	agentConfigLocation = "/config/metric-collector.yaml"
+)
+
+type Configurator struct {
+	TelemetrySpec telemetryapi.TelemetrySpec
+	Kluster       *k8ssandraapi.K8ssandraCluster
+	Ctx           context.Context
+	RemoteClient  client.Client
+}
+
+func (c Configurator) GetTelemetryAgentConfigMap() (corev1.ConfigMap, error) {
+	yamlData, err := yaml.Marshal(&c.TelemetrySpec.Cassandra)
+	if err != nil {
+		return corev1.ConfigMap{}, err
+	}
+	cm := corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: c.Kluster.Namespace,
+			Name:      c.Kluster.Name + "metrics-agent-config",
+		},
+		Data: map[string]string{filepath.Base(agentConfigLocation): string(yamlData)},
+	}
+	return cm, nil
+}
+
+func (c Configurator) ReconcileTelemetryAgentConfig(dc *cassdcapi.CassandraDatacenter) error {
+	cm, err := c.GetTelemetryAgentConfigMap()
+	if err != nil {
+		return err
+	}
+	if err := c.RemoteClient.Create(c.Ctx, &cm); err != nil {
+		return err
+	}
+	c.AddStsVolumes(dc)
+
+	return nil
+}
+
+func (c Configurator) AddStsVolumes(dc *cassdcapi.CassandraDatacenter) error {
```

Comment:
Could you reuse `AddOrUpdateVolume()` which is in `pkg/cassandra/datacenter.go` ?

Reply:
I had a close look at this and wasn't quite sure that it was appropriate. I'm open minded here, but it seems that the function you're referring to works on the `datacenterConfig` type, whereas we currently work directly on the cassDC and hook into the reconciler at that level. 

If you want me to investigate a change of this nature in more detail, I'm happy to do so, but it is a fairly deep re-write.

Reply:
The advantage is that using that function does overwrite any existing `Volume` which might already exist. But also, we don't usually provide guarantees in our controllers that a manually modified object will be self-healed back to the desired state. This is in line with in-built k8s controllers like the STS controller.